core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardsStoresAction.java

```diff
@@ -0,0 +1,255 @@
+                    InternalAsyncFetch fetch = new InternalAsyncFetch(logger, "shard_stores", shardId, listShardStoresInfo);
+                    fetch.fetchData(nodes, metaData, Collections.<String>emptySet());
+                }
+            }
+        }
+
+        private class InternalAsyncFetch extends AsyncShardFetch<NodeGatewayStartedShards> {
+
+            InternalAsyncFetch(ESLogger logger, String type, ShardId shardId, TransportNodesListGatewayStartedShards action) {
+                super(logger, type, shardId, action);
+            }
+
+            @Override
+            protected synchronized void processAsyncFetch(ShardId shardId, NodeGatewayStartedShards[] responses, FailedNodeException[] failures) {
+                fetchResponses.add(new Response(shardId, responses, failures));
+                if (expectedOps.countDown()) {
+                    finish();
+                }
+            }
+
+            void finish() {
+                Map<String, IntObjectHashMap<java.util.List<NodeGatewayStartedShards>>> shardsResponses = new HashMap<>();
+                ImmutableList.Builder<IndicesShardsStoresResponse.Failure> failures = ImmutableList.builder();
+                while (!fetchResponses.isEmpty()) {
+                    Response res = fetchResponses.remove();
+                    IntObjectHashMap<java.util.List<NodeGatewayStartedShards>> indexShards = shardsResponses.get(res.shardId.getIndex());
+                    final IntObjectHashMap<java.util.List<NodeGatewayStartedShards>> indexShardsBuilder;
+                    if (indexShards == null) {
+                        indexShardsBuilder = new IntObjectHashMap<>();
+                    } else {
+                        indexShardsBuilder = new IntObjectHashMap<>(indexShards);
+                    }
+                    java.util.List<NodeGatewayStartedShards> shardStoreStatuses = indexShardsBuilder.get(res.shardId.id());
+                    if (shardStoreStatuses == null) {
+                        shardStoreStatuses = new ArrayList<>();
+                    }
+                    for (NodeGatewayStartedShards response : res.responses) {
+                        if (shardExistsInNode(response)) {
+                            shardStoreStatuses.add(response);
+                        }
+                    }
+                    indexShardsBuilder.put(res.shardId.id(), shardStoreStatuses);
+                    shardsResponses.put(res.shardId.getIndex(), indexShardsBuilder);
+                    for (FailedNodeException failure : res.failures) {
+                        failures.add(new IndicesShardsStoresResponse.Failure(failure.nodeId(), res.shardId.getIndex(), res.shardId.id(), failure.getCause()));
+                    }
+                }
+                ImmutableOpenMap.Builder<String, ImmutableOpenIntMap<java.util.List<IndicesShardsStoresResponse.ShardStoreStatus>>> indicesShardStoreStatuses = ImmutableOpenMap.builder();
```

Comment:
this confuses me.. why does it need to be a separate loop? can we inline it with the previous one?